### PR TITLE
AC Event Handling

### LINKED ISSUE #8: Automatically enter High Performance when plugged in
When connected to power (and perhaps once the battery is charged to a certain percent), we should switch into High Performance mode. I don't see a strong argument for trying to save power when connected to an effectively unlimited source.

Of course we should switch back to whichever profile the user had set if they disconnect power.

### LINKED ISSUE #9: Automatically enter Battery Life when battery is low
Inversely to #8, we should consider dropping into Battery Life mode when the battery is below a certain percentage and the device is running on battery power. Like the other modes, the user should be able to toggle it back to whichever profile they'd like.

### LINKED ISSUE #21: Automatically enter Balanced when disconnecting from power
Specified in #8, but more explicitly stated here. When a laptop disconnects from power, we should enter Balanced mode (unless the battery is low).

### LINKED ISSUE #38: Power settings lack performance option
**NAME="Pop!_OS"
VERSION="18.04 LTS"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Pop!_OS 18.04 LTS"
VERSION_ID="18.04"
HOME_URL="https://system76.com/pop"
SUPPORT_URL="http://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
**

**Issue/Bug Description:**
There is no way to set a performance mode. The only way to set is through the current **System76 Power extension** between:
- High Performance
- Balanced
- Battery Life

Nevertheless, it is currently **impossible** to change this extension's default to high performance when connected to electricity.

**Steps to reproduce (if you know):**
A. Settings
- Access Settings > Power
- Fail to change efficiency mode 

B. System76 Power extension
- Access extension's drop-down menu
- Change to a different value. E.G.: High performance
- Restart computer
- default goes back to Balanced

**Expected behavior:**
A. Be able to set an efficiency mode
B. Be able to set a new default

### PR DESCRIPTION
**Pull request by mmstick**
**Tested features**

- [x] If on AC, and the user switches to battery, the balanced profile will be set.
- [x] If on battery, and the user switches to AC, the performance profile will be set.
- [x] If on battery, and the battery drops below 25% (critical), the battery profile will be set.
- [x] If on battery, and the battery rises above 50% after previously reaching critical, the balanced profile will be set.

Closes #8
Closes #9 
Closes #21 
Closes #38 

Requires #42 to make the settings configurable

### CODE REVIEW

**Comment by mmstick** [comment]
Finally found a solution to the build error here. We're going to need to create a compressed tarball of our vendored sources that preserves permissions outside of the schroot, so that `dpkg-buildpackage` doesn't tamper with them when it creates the source tarball. The fix to the Makefile will be pushed soon.

**Comment by jackpot51** [comment]
Interesting

File: Makefile [hunk 1]
```diff
@@ -35,12 +35,17 @@ update:
 	cp $< $@
 
 vendor: .cargo/config
-	cargo vendor
-	touch vendor
+	if [ ! -f vendor.tar.xz ]; then \
+		cargo vendor; \
+		touch vendor; \
+		tar pcfJ vendor.tar.xz vendor; \
+		rm vendor -r; \
```
Comment (jackpot51) [inline Makefile]: I prefer `rm -r vendor`.

File: Makefile [hunk 2]
```diff
@@ -35,12 +35,17 @@ update:
 	cp $< $@
 
 vendor: .cargo/config
-	cargo vendor
-	touch vendor
+	if [ ! -f vendor.tar.xz ]; then \
+		cargo vendor; \
+		touch vendor; \
```
Comment (jackpot51) [inline Makefile]: touch vendor is not required

File: Makefile [hunk 3]
```diff
@@ -35,12 +35,17 @@ update:
 	cp $< $@
 
 vendor: .cargo/config
```
Comment (jackpot51) [inline Makefile]: the target should be vendor.tar.xz

File: Makefile [hunk 4]
```diff
@@ -35,12 +35,17 @@ update:
...
-	cargo vendor
-	touch vendor
+	if [ ! -f vendor.tar.xz ]; then \
```
Comment (jackpot51) [inline Makefile]: if you use the target as vendor.tar.xz, you do not need the if statement

**Comment by brs17** [comment]
I am hesitant to go with these defaults...in discussing this pr in the office, we came up with a number of edge cases where this functionality is potentially not what a user would want. 

I definitely like the idea for the user to be able to set these types of settings but I am not sure these defaults are what we would want to offer out of the box.

**Comment by mmstick** [comment]
Deferring additional work on this until #42 is merged.

**Comment by andirsun** [comment]
@mmstick #42 is closed so I really need this PR, would be possible move forward with this PR?

**Comment by dhpowrhost** [comment]
This feature would really be nice.

**Pull request closed without merging**